pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/generator/BaseTaskGenerator.java

```diff
@@ -117,6 +118,26 @@ public List<SegmentZKMetadata> getSegmentsZKMetadataForTable(String tableNameWit
     return selectedSegmentZKMetadataList;
   }
 
+  public List<SegmentZKMetadata> getNonConsumingSegmentsZKMetadataForRealtimeTable(String tableNameWithType) {
+    IdealState idealState = _clusterInfoAccessor.getIdealState(tableNameWithType);
+    Set<String> idealStateSegments = idealState.getPartitionSet();
+    List<SegmentZKMetadata> segmentZKMetadataList = _clusterInfoAccessor.getSegmentsZKMetadata(tableNameWithType);
```

Comment:
This is a super expensive operation for large table and should be prevented.

We should change the order to first gather segments in CONSUMING state, then only read ZK metadata for them.

Does it work if we change the validation manager to also fix the UPLOADED segment when it is in CONSUMING state? That is much easier to manage than forcing all tasks to apply this check